### PR TITLE
Dash oil gas ternary

### PR DESCRIPTION
**Pull request by YunkeXiao**
Issue for app: #60

# App pull request
- [ ] This is a new app
- [x] I am improving an existing app (redesigns/code "makeovers")

## About

- Playground deployment URL (new version):https://dash-playground.plotly.host/dash-oil-gas-ternary/

## Workflow

- [x] I have created a branch in the appropriate monorepo, and the
  elements necessary for successful deployment are in place.
- [x] If the app is a redesigned and/or restyled version of an
  existing gallery app, I've summarized the changes requested in the
  appropriate Streambed issue and confirm that they have been applied.
- [x] If the app is on the Dash Gallery portal, I have added a link to
  the GitHub repository for the source code in the portal description.
- [x] If the app is a reimplementation of a Python gallery app for the
  DashR gallery, the app in this PR mimics, as closely as possible,
  the style and functionality of the existing app.

## The pre-review review

I have addressed all of the following questions:

- [x] Does everything in my code serve some purpose? (I have removed
  any dead and/or irrelevant code.)
- [x] Does everything in my code have a clear purpose? (My code is
  readable and, where it isn't, it has been commented appropriately.)]
- [x] Am I reinventing the wheel? (I have used appropriate packages to
  lessen the volume of code that needs to be maintained.)

### CODE REVIEW

File: apps/dash-oil-gas-ternary/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+## Dash-oil-gas-ternary
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/README.md]:
```suggestion
## dash-oil-gas-ternary
```

File: apps/dash-oil-gas-ternary/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+`dash-oil-gas-ternary` creates a dashboard for mineral composition evaluations from natural gas wells.  
+By pairing geographic map of well locations with ternary diagram, this app helps geologists to find out results of gas wells with relationship to formations or rock types.
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/README.md]:
```suggestion
By pairing a geographic map of well locations with a ternary diagram, this app helps geologists visualize gas wells and the constituent minerals of the earth that surrounds them.
```

File: apps/dash-oil-gas-ternary/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+and all of the required `pip` packages, will be installed, and the app will be able to run.
+
+## How to use the app
+Run this app locally by:
+```
+python app.py
+```
+Open http://127.0.0.1:8050/ in your browser.
+
+Select data points from well map, ternary map or bar graph to visualize cross-filtering to other plots.
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/README.md]:
```suggestion
Select a subset of data points from the well map, ternary map or bar graph to visualize cross-filtering to other plots.
```

File: apps/dash-oil-gas-ternary/README.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+Selection could be done by clicking on individual data points or using the lasso tool to capture multiple data points or bars. Hold SHIFT
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/README.md]:
```suggestion
Selection can be done by clicking on individual data points, or by using the lasso tool to capture multiple data points or bars. Hold the SHIFT
```

File: apps/dash-oil-gas-ternary/README.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+button while click-and-drag for multi-region selection.
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/README.md]:
```suggestion
key while clicking and dragging for multi-region selection.
```

File: apps/dash-oil-gas-ternary/app.py [hunk 6]
```diff
@@ -0,0 +1,791 @@
...
+            html.H6("Oil and gas ternary map"),
+        ],
+    )
+
+
+def build_graph_title(title):
+    return html.P(className="graph-title", children=title)
+
+
+def generate_production_plot(processed_data):
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/app.py]: If the function below has a docstring (which is awesome!), please add docstrings to the other ones as well! :)

File: apps/dash-oil-gas-ternary/app.py [hunk 7]
```diff
@@ -0,0 +1,791 @@
...
+                    children=[
+                        html.Div(
+                            className="six columns",
+                            id="header-container",
+                            children=[
+                                build_banner(),
+                                html.P(
+                                    id="instructions",
+                                    children="Select data points from the well map, ternary map or bar graph to visualize cross-filtering to"
+                                    " other plots. Selection could be done by clicking on individual data points or using the lasso tool "
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/app.py]:
```suggestion
                                    " other plots. Selection could be done by clicking on individual data points or by using the lasso tool "
```

File: apps/dash-oil-gas-ternary/app.py [hunk 8]
```diff
@@ -0,0 +1,791 @@
...
+                                    "to capture multiple data points or bars. With the box tool from modebar, multiple regions can be selected by holding SHIFT "
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/app.py]:
```suggestion
                                    "to capture multiple data points or bars. With the box tool from modebar, multiple regions can be selected by holding the SHIFT key"
```

File: apps/dash-oil-gas-ternary/app.py [hunk 9]
```diff
@@ -0,0 +1,791 @@
...
+                                    "button while click-and-drag.",
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/app.py]:
```suggestion
                                    "while clicking and dragging.",
```

File: apps/dash-oil-gas-ternary/app.py [hunk 10]
```diff
@@ -0,0 +1,791 @@
...
+    else:
+        for formation in formations:
+            processed_data[formation] = None
+
+    return generate_well_map(dff, processed_data, mapbox_view)
+
+
+# Update production plot
+@app.callback(
+    output=Output("production-fig", "figure"),
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/app.py]: We actually don't typically include these keywords in the arguments of a callback. 
```suggestion
    Output("production-fig", "figure"),
```

File: apps/dash-oil-gas-ternary/app.py [hunk 11]
```diff
@@ -0,0 +1,791 @@
...
+    inputs=[
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/app.py]:
```suggestion
    [
```

File: apps/dash-oil-gas-ternary/requirements.txt [hunk 12]
```diff
@@ -0,0 +1,5 @@
+dash>=0.41.0
```
Comment (shammamah) [inline apps/dash-oil-gas-ternary/requirements.txt]: Please update to the latest version of Dash!

File: apps/dash-oil-gas-ternary/README.md [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+![Animated](assets/Screencast.gif)
+
+## Screenshots
+![initial](assets/Screenshot.png)
+
+## Requirements
+We suggest you to create a separate virtual environment running Python 3 for this app, and install all of the required dependencies there. Run in Terminal/Command Prompt:
+
+```
+git clone https://github.com/plotly/dash-oil-gas-ternary.git
```
Comment (celinehuang) [inline apps/dash-oil-gas-ternary/README.md]: @shammamah told me these repo will be archived so probably put the monorepo link here and have them cd into dash-sample-apps/apps/dash-oil-gas-ternary

Comment (ycaokris) [reply apps/dash-oil-gas-ternary/README.md]: I'd add a migration/deprecated notice in original repo readme.

**Comment by shammamah** [comment]
@YunkeXiao Could you try to keep a max height for the `dcc.Dropdown`? You can check this out for some inspiration: https://github.com/plotly/dash-bio/blob/d16a10d7a4444c37b8da3a515171100594b46b20/assets/clustergram-style.css#L126

File: apps/dash-oil-gas-ternary/app.py [hunk 14]
```diff
@@ -0,0 +1,790 @@
...
+    colormap[formation_name] = constants.colors[ind]
+
+
+def build_banner():
+    return html.Div(
+        id="banner",
+        className="banner",
+        children=[
+            html.Img(
+                src="https://files.slack.com/files-pri/T06LPNGUD-FKXDDD4QG/dash-logo-new.png"
```
Comment (celinehuang) [inline apps/dash-oil-gas-ternary/app.py]: image
Logo doesn't show up, I think it's better to download it and put it into the assets folder and `src=app.get_asset_url("dash-logo-new.png")`

**Comment by celinehuang** [comment]
Some things I noticed with the app - 
1. weird spacing with the dropdown
image
image

2. maybe space out the radio buttons/ check boxes, and make sure the mode bar isn't covering the legend 
image

everything else looks good! just some small fixes :)

**Comment by ycaokris** [comment]
Some minor suggestions:

* Composition label(Quartz, Carbonate, Clay) for ternary app is missing, could be partially hidden by ternary plot sizing.  it may be helpful to display it back. 
* I didn't see annotations on well-map, since well-category is already reflected by colors/ axis labels from bar plot, just remove this annotation if positioning is difficult.

* The top two plots (well-map and ternary map) are not initialized in layout,  and generated by callback firing at page load, since firing callback happens after layout loading, you will see an empty figure with white paper background. This could be resolved by configuring empty dcc.Graph in app.layout [here](https://github.com/plotly/dash-sample-apps/blob/2eb0bd109cfd0e5827aaaf0881f58c310fa87735/apps/dash-oil-gas-ternary/app.py#L437) and [here](https://github.com/plotly/dash-sample-apps/blob/2eb0bd109cfd0e5827aaaf0881f58c310fa87735/apps/dash-oil-gas-ternary/app.py#L469) like : 
```
figure={'data': [], 'layout': {"paper_bgcolor": "#192444",
        "plot_bgcolor": "#192444"}}
```
for a smoother transit

* It seems graphs appear overflow when I zoom 100% at 1400px. 

screencapture-dash-playground-plotly-host-dash-oil-gas-ternary-2019-07-09-14_10_14

File: apps/dash-oil-gas-ternary/app.py [hunk 15]
```diff
@@ -0,0 +1,790 @@
...
+for ind, formation_name in enumerate(df["fm_name"].unique().tolist()):
+    colormap[formation_name] = constants.colors[ind]
+
+
+def build_banner():
+    return html.Div(
+        id="banner",
+        className="banner",
+        children=[
+            html.Img(src=app.get_asset_url("dash-logo.png")),
```
Comment (celinehuang) [inline apps/dash-oil-gas-ternary/app.py]: I personally think the dash logo is a little big too big for the desktop version compare to the mockup, but that's up to your discretion

**Comment by YunkeXiao** [comment]
> Screen Shot 2019-07-18 at 1 29 01 PM
> 
> Please make sure this is centered in the mobile layout. Otherwise, 💃

There was no mock-up for the mobile  version so I took the liberty of making my own design choices and I personally think that it made more sense, on my phone, for the checkboxes not to be centered.

File: apps/dash-oil-gas-ternary/requirements.txt [hunk 16]
```diff
@@ -0,0 +1,5 @@
+dash>=0.43.0
+dash-daq>=0.1.5
```
Comment (ycaokris) [inline apps/dash-oil-gas-ternary/requirements.txt]:
```suggestion

```

File: apps/dash-oil-gas-ternary/app.py (hunk 6 above)
Comment (ycaokris) [reply apps/dash-oil-gas-ternary/app.py]: please add docstring for this function

File: apps/dash-oil-gas-ternary/app.py [hunk 17]
```diff
@@ -0,0 +1,803 @@
...
+        margin=dict(l=0, r=0, t=0, b=0),
+        mapbox=go.layout.Mapbox(
+            accesstoken=mapbox_access_token,
+            bearing=0,
+            center=go.layout.mapbox.Center(lat=37.497562, lon=-82.755728),
+            pitch=0,
+            zoom=8,
+            style=style,
+        ),
+        legend=dict(
```
Comment (ycaokris) [inline apps/dash-oil-gas-ternary/app.py]: nice workaround for positioning legend 👍

**Pull request merged**